### PR TITLE
SubType feature part I - Enable subtype read for projection

### PR DESCRIPTION
**Pull request by YannanGao-gs**
## Summary
Partially fixes https://github.com/finos/legend-studio/issues/630
SubType feature part I - Enable subtype read for projection

## How did you test this change?

- [ ] Test(s) added
- [x] Manual testing (please provide screenshots/recordings)
- [ ] No testing (please provide an explanation)

https://user-images.githubusercontent.com/73408381/146020896-c6d4d094-91c2-4553-81a6-f4330ed13e39.mov

https://user-images.githubusercontent.com/73408381/146020930-2fc01495-58a6-4808-b02a-6ef673f3034c.mov

### CODE REVIEW

File: packages/legend-graph/src/models/protocols/pure/v1/transformation/pureGraph/to/helpers/V1_ValueSpecificationBuilderHelper.ts [hunk 1]
```diff
@@ -869,6 +872,16 @@ export function V1_processProperty(
...
+  if (
+    inferredVariable instanceof SimpleFunctionExpression &&
+    matchFunctionName(
+      inferredVariable.functionName,
+      'meta::pure::functions::lang::subType',
+    )
+  ) {
+    inferredType =
+      inferredVariable.parametersValues[1]?.genericType?.value.rawType;
+  }
```
Comment (MauricioUyaguari) [inline packages/legend-graph/src/models/protocols/pure/v1/transformation/pureGraph/to/helpers/V1_ValueSpecificationBuilderHelper.ts:884]: Roundtrip Lamba Test ?

File: .changeset/perfect-zebras-crash.md [hunk 2]
```diff
@@ -0,0 +1,6 @@
+---
+'@finos/legend-graph': patch
+'@finos/legend-query': patch
+---
+
+Enabled read process of subtype for projection
```
Comment (MauricioUyaguari) [inline .changeset/perfect-zebras-crash.md:6]: `Add processing support for subtype in projection fetch structure`.

**Comment by MauricioUyaguari** [comment]
We need RoundTrip Tests. 
`QueryBuilder_LambdaProcessingRoundtrip.test` and `QueryBuilder_BuildLambdaRoundtrip.test`

**Review by MauricioUyaguari — CHANGES_REQUESTED** [review]
Seems like we are missing `QueryBuilderLambdaBuilder` changes for the roundtrip to work.

File: packages/legend-graph/src/models/protocols/pure/v1/transformation/pureGraph/to/helpers/V1_ValueSpecificationBuilderHelper.ts (hunk 1 above)
Comment (YannanGao-gs) [reply packages/legend-graph/src/models/protocols/pure/v1/transformation/pureGraph/to/helpers/V1_ValueSpecificationBuilderHelper.ts:884]: As we discussed offline, I tested locally to make sure that added tests do go through the processors by writing some console.log statements.
(heads up: even though there is a typo or only tests themself without any necessary changes, tests still pass)

File: packages/legend-query/src/stores/QueryBuilderPropertyEditorState.ts [hunk 3]
```diff
@@ -58,6 +61,24 @@ export const getPropertyChainName = (
...
+        SUPPORTED_FUNCTIONS.SUBTYPE,
+      )
+    ) {
+      const propertyName =
+        currentExpression.parametersValues[0] instanceof
+        AbstractPropertyExpression
+          ? currentExpression.parametersValues[0]?.func.name
+          : '';
+      const propertyWithSubtype = `(@${currentExpression.parametersValues[1]?.genericType?.value.rawType.name}) ${propertyName}`;
+      propertyNameChain.unshift(propertyWithSubtype);
```
Comment (akphi) [inline packages/legend-query/src/stores/QueryBuilderPropertyEditorState.ts:77]: let's inline `propertyWithSubtype`

**Review by akphi — CHANGES_REQUESTED** [review]

File: .changeset/grumpy-bees-study.md [hunk 4]
```diff
@@ -0,0 +1,6 @@
+---
+'@finos/legend-graph': patch
+'@finos/legend-query': patch
```
[the remark this answers is not in the document]
Comment (akphi) [reply .changeset/grumpy-bees-study.md:3]: also, these changes seem like `minor` to me :)

File: packages/legend-graph/src/models/protocols/pure/v1/transformation/pureGraph/to/helpers/V1_ValueSpecificationBuilderHelper.ts [hunk 5]
```diff
@@ -871,6 +872,16 @@ export function V1_processProperty(
   if (inferredVariable instanceof AbstractPropertyExpression) {
     inferredType = inferredVariable.func.genericType.value.rawType;
   }
+  if (
+    inferredVariable instanceof SimpleFunctionExpression &&
+    matchFunctionName(
+      inferredVariable.functionName,
+      'meta::pure::functions::lang::subType',
```
[the remark this answers is not in the document]
Comment (MauricioUyaguari) [reply packages/legend-graph/src/models/protocols/pure/v1/transformation/pureGraph/to/helpers/V1_ValueSpecificationBuilderHelper.ts:879]: I think we should do this. @YannanGao-gs can you do this in a separate PR after we get this merged. Thanks

[the remark this answers is not in the document]
Comment (akphi) [reply packages/legend-graph/src/models/protocols/pure/v1/transformation/pureGraph/to/helpers/V1_ValueSpecificationBuilderHelper.ts:879]: why do we need to do this in a separate PR?

File: packages/legend-query/src/components/__tests__/QueryBuilder_FetchStructure.test.tsx [hunk 6]
```diff
@@ -148,6 +149,34 @@ test(
     expect(lastNameProperty).toBe(_personClass.getProperty('lastName'));
     expect(queryBuilderState.resultSetModifierState.limit).toBeUndefined();
 
+    // simpleProjection with subType
```
Comment (akphi) [inline packages/legend-query/src/components/__tests__/QueryBuilder_FetchStructure.test.tsx:152]: @MauricioUyaguari this 200-line compound test is freaking me out. Should we split this up?

File: packages/legend-query/src/stores/QueryBuilderPropertyEditorState.ts [hunk 7]
```diff
@@ -58,6 +61,25 @@ export const getPropertyChainName = (
...
       propertyNameChain.unshift(currentExpression.func.name);
     }
+    if (
+      currentExpression instanceof SimpleFunctionExpression &&
+      matchFunctionName(
+        currentExpression.functionName,
+        SUPPORTED_FUNCTIONS.SUBTYPE,
+      )
+    ) {
+      const propertyName =
```
Comment (akphi) [inline packages/legend-query/src/stores/QueryBuilderPropertyEditorState.ts:71]: looks like `propertyName` is nullable

**Review by akphi — CHANGES_REQUESTED** [review]

File: packages/legend-query/src/stores/QueryBuilderPropertyEditorState.ts (hunk 7 above)
Comment (YannanGao-gs) [reply packages/legend-query/src/stores/QueryBuilderPropertyEditorState.ts]: yeah you are right! sry  fixed this.

File: packages/legend-query/src/components/__tests__/QueryBuilder_FetchStructure.test.tsx (hunk 6 above)
Comment (MauricioUyaguari) [reply packages/legend-query/src/components/__tests__/QueryBuilder_FetchStructure.test.tsx:152]: Yes. @YannanGao-gs. Lets split these out. Each time we initialize queryBuilderState we should have a new test.

File: packages/legend-graph/src/models/protocols/pure/PureProtocolProcessorPlugin.ts [hunk 8]
```diff
@@ -139,4 +140,11 @@ export abstract class PureProtocolProcessorPlugin extends AbstractPlugin {
    * This would provide a mechanism to add more elements in this reduced graph.
    */
   V1_getExtraExecutionInputGetters?(): V1_ExecutionInputGetter[];
+
+  /**
+   * Get the type of the inferred variable in property expression.
+   */
+  V1_getExtraPropertyExpressionTypeInferenceProcedures?(
```
Comment (MauricioUyaguari) [inline packages/legend-graph/src/models/protocols/pure/PureProtocolProcessorPlugin.ts:147]: Lets follow the patterns of the other extensions. You should return a list of functions you can call that will return Type|undefined. Look at `V1_FunctionExpressionBuilder`

File: .changeset/grumpy-bees-study.md [hunk 9]
```diff
@@ -0,0 +1,5 @@
+---
+'@finos/legend-graph': minor
+---
+
+Handle subtype in property processing.
```
Comment (MauricioUyaguari) [inline .changeset/grumpy-bees-study.md:5]: This changeset note is for `legend-graph`. Subtype should not be mentioned at all. This changeset should address the extension you are adding.

File: packages/legend-graph/src/models/protocols/pure/v1/transformation/pureGraph/to/helpers/V1_ValueSpecificationBuilderHelper.ts [hunk 10]
```diff
@@ -868,9 +868,10 @@ export function V1_processProperty(
   }
   let inferredType: Type | undefined =
     inferredVariable?.genericType?.value.rawType;
-  if (inferredVariable instanceof AbstractPropertyExpression) {
```
Comment (MauricioUyaguari) [inline packages/legend-graph/src/models/protocols/pure/v1/transformation/pureGraph/to/helpers/V1_ValueSpecificationBuilderHelper.ts:871]: Only the subtype condition should be moved to the plugin. The `(inferredVariable instanceof AbstractPropertyExpression` should remain in core.

**Review by MauricioUyaguari — APPROVED** [review]

**Pull request merged**